Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/security/EncryptionErrorResetTool.kt

```diff
@@ -0,0 +1,137 @@
+package de.rki.coronawarnapp.util.security
+
+import android.content.Context
+import android.content.SharedPreferences
+import androidx.core.content.edit
+import de.rki.coronawarnapp.storage.DATABASE_NAME
+import de.rki.coronawarnapp.util.errors.causes
+import org.joda.time.Instant
+import timber.log.Timber
+import java.io.File
+import java.security.GeneralSecurityException
+import javax.inject.Inject
+import javax.inject.Singleton
+
+/**
+ * This tool determines the narrow scope for which we will recovery from an encryption error
+ * by resetting our encrypted data.
+ * This will allow users currently affected by it, that update the app, to keep using it without
+ * requiring any manual actions from their side.
+ *
+ * https://github.com/corona-warn-app/cwa-app-android/issues/642
+ */
+@Singleton
+class EncryptionErrorResetTool @Inject constructor(
+    private val context: Context
+) {
+    // https://cs.android.com/android/platform/superproject/+/master:frameworks/base/core/java/android/app/ContextImpl.java;drc=3b8e8d76315f6718a982d5e6a019b3aa4f634bcd;l=626
+    private val encryptedPreferencesFile by lazy {
+        val appbaseDir = context.filesDir.parentFile!!
+        val sharedPrefsDir = File(appbaseDir, "shared_prefs")
+        File(sharedPrefsDir, "${SecurityConstants.ENCRYPTED_SHARED_PREFERENCES_FILE}.xml")
+    }
+    private val encryptedDatabaseFile by lazy {
+        context.getDatabasePath(DATABASE_NAME)
+    }
+    private val prefs: SharedPreferences by lazy {
+        context.getSharedPreferences("encryption_error_reset_tool", Context.MODE_PRIVATE)
+    }
+
+    private var isResetWindowConsumed: Boolean
+        get() = prefs.getBoolean(PKEY_EA1851_WAS_WINDOW_CONSUMED, false)
+        set(value) = prefs.edit {
+            putBoolean(PKEY_EA1851_WAS_WINDOW_CONSUMED, value)
+        }
+
+    private var resetPerformedAt: Instant?
+        get() {
+            val performedAt = prefs.getLong(PKEY_EA1851_RESET_PERFORMED_AT, -1)
+            return if (performedAt == -1L) null else Instant.ofEpochMilli(performedAt)
```

Comment:
define const INVALID = -1L to make it more clear?

Reply:
Not here, it's so close to each other. I would have also used `null` for the check but `getLong` doesn't support `Long?` as default value.